src/parser/shared/metrics/buffApplications.ts

```diff
@@ -0,0 +1,82 @@
+import { AnyEvent, EventType } from 'parser/core/Events';
+import metric from 'parser/core/metric';
+
+/**
+ * Returns the index of the last element in the array where predicate is true, and -1
+ * otherwise.
+ * Source: https://stackoverflow.com/a/53187807/684353
+ * @param array The source array to search in
+ * @param predicate find calls predicate once for each element of the array, in descending
+ * order, until it finds one where predicate returns true. If such an element is found,
+ * findLast immediately returns that element index. Otherwise, findLastIndex returns undefined.
+ */
+export function findLast<T>(
+  array: T[],
+  predicate: (value: T, index: number, obj: T[]) => boolean,
+): T | undefined {
+  let l = array.length;
+  while (l !== 0) {
+    l -= 1;
+    if (predicate(array[l], l, array)) {
+      return array[l];
+    }
+  }
+  return undefined;
+}
+
+/**
+ * Returns an object with all buff applications. Should probably not be used
+ * directly in configs, but only internally by other shared stats.
+ */
+const buffApplications = (events: AnyEvent[]) =>
+  events.reduce<{
+    [spellId: number]: {
+      [sourceId: number]: Array<{
+        targetId: number;
+        targetInstance?: number;
+        start?: number;
+        end?: number;
```

Comment:
I could also see myself adding refreshTimes to this as well, but that can be a future upgrade

Reply:
Might also be a separate metric.